### DISCUSSION TITLE
Best way to overload a dataset based on the data type of an input?

### DISCUSSION THREAD

**Discussion by richardhathaway-8451** [8451/labrea#23] [Q&A] [answered]
@austinwarner-8451  - I was wondering if there is an easy way to overload a dataset or abstractdataset based on the data type of an input? This is the best strategy I could come up with, but maybe I've missed something!

```
from labrea import abstractdataset, Option, Switch

@abstractdataset(
    dispatch = Switch(
        Option("X") >> (lambda x: type(x)),
        {
            str: "string",
            int: "int",
            list: "list"
        },
        "something else" 
    )
)
def foo():
    ...

@foo.overload(alias = "string")
def foo_str():
    return "I'm a string!"
  
@foo.overload(alias = "int")
def foo_int():
    return "I'm an int!"
  
@foo.overload(alias = "list")
def foo_list():
    return "I'm a list!"
  
@foo.overload(alias = "something_else")
def foo_somethingelse():
    return "I'm something else!"

```

**Comment by richardhathaway-8451** [comment, marked as answer]
Actually I don't think the Switch is necessary, I can just do this:

```
@dataset(
    dispatch = (
        Option("X") >> (lambda x: type(x))
    )
)
def foo():
    return "I'm something else!"

@foo.overload(alias = str)
def foo_str():
    return "I'm a string!"
  
@foo.overload(alias = int)
def foo_int():
    return "I'm an int!"
  
@foo.overload(alias = list)
def foo_list():
    return "I'm a list!"
```

**Reply by austinwarner-8451** [reply]
This is the current best option (though technically `lambda x: type(x)` is redundant, you could just do `Option("X") >> type`)

The one downside to this is type inheritance will break this. For example

```python
class Fruit:
    pass

class Apple(Fruit):
    pass

class Gala(Apple):
    pass

@dataset
def fruit() -> Fruit:
    return Gala()

@abstractdataset(dispatch=fruit >> type)
def fruit_type() -> str:
    pass

@name.overload(alias=Apple)
def fruit_type_apple():
    return 'Apple'

fruit_type()  
# SwitchError: Evaluated to <class Gala>, but must be one of <class Apple>
```

Labrea does expose a [case/when construct](https://8451.github.io/labrea/api.html#labrea.case), but overloads only work with switches under the hood. I have toyed around in the past with adding support for registering an overload using a case in a case/when clause, in addition to the alias used by a switch. That would be a potentially good fit here if the types you need to handle exist in a hierarchy and you want a subclass to still be captured (using an isisntance check in the case)
